### PR TITLE
feat(bigtable): Support non-String primary key types for row keys

### PR DESCRIPTION
**Pull request by wiliam-oliveira**
## Summary

- Relax the row key type constraint from `VARCHAR`/`CHAR`-only to also accept `BIGINT`, `INTEGER`, `SMALLINT`, and `TINYINT`
- Numeric types are zero-padded to 19 digits at serialization time to preserve lexicographic sort order for non-negative values, as recommended by the Bigtable schema design guide
- Update README and Javadoc to reflect the new supported row key types

## Motivation

The Bigtable connector currently requires the primary key (row key) to be of type `VARCHAR`/`CHAR`. This forces users whose source data has numeric primary keys (e.g., `INT`, `BIGINT`) to either:

1. Insert an explicit `CAST(id AS STRING)` in their SQL pipeline, or
2. Accept that Flink's planner will automatically insert a `SinkMaterializer` to coerce the type, adding an unnecessary materialization step to the pipeline

Both workarounds add unnecessary complexity. For example, consider a pipeline that writes product data to Bigtable where the source has a `BIGINT` primary key:

```sql
-- Source table with a BIGINT primary key
CREATE TABLE product_changes (
    product_id BIGINT PRIMARY KEY NOT ENFORCED,
    shop_id    BIGINT,
    title      STRING
) WITH (
    'connector' = 'kafka',
    ...
);

-- Sink table
CREATE TABLE products_bigtable (
    product_id BIGINT PRIMARY KEY NOT ENFORCED,
    shop_id    BIGINT,
    title      STRING
) WITH (
    'connector' = 'bigtable',
    'project'   = 'my-project',
    'instance'  = 'my-instance',
    'table'     = 'products',
    'column-family' = 'product'
);

-- Without this patch: fails because the connector rejects BIGINT PKs, forcing you to cast
-- and redefine the sink schema as STRING. The cast also causes Flink's planner to insert
-- a SinkMaterializer, like the one in picture below, adding an unnecessary materialization
-- step to the pipeline.
INSERT INTO products_bigtable
SELECT CAST(product_id AS STRING), shop_id, title
FROM product_changes;
```

image

With this patch: works directly, no cast needed

```sql
INSERT INTO products_bigtable
SELECT product_id, shop_id, title
FROM product_changes;
```

image

Since Bigtable row keys are ultimately byte strings, and the conversion from numeric types to zero-padded strings is deterministic, lossless, and preserves lexicographic sort order for non-negative values, the connector should handle this natively.

### Source

| File | Change |
|---|---|
| `BigtableDynamicTableSink.java` | Remove duplicate `SUPPORTED_ROW_KEY_TYPES`; delegate to `RowDataToRowMutationSerializer.SUPPORTED_ROW_KEY_TYPES` |
| `RowDataToRowMutationSerializer.java` | Make `SUPPORTED_ROW_KEY_TYPES` public; add `extractRowKeyAsString()` with zero-padded numeric formatting; widen type check in `generateMapsFromSchema()`; store `rowKeyTypeRoot` |
| `ErrorMessages.java` | Use correct `LogicalTypeRoot` names (`VARCHAR`, `CHAR`) instead of `STRING` in the error template |

### Tests

| File | Change |
|---|---|
| `BigtableTableTest.java` | Add `testNonStringPrimaryKeyAccepted` (parameterized for all types), `testUnsupportedRowKeyType` (DOUBLE), update `rowKeyCases` |
| `RowDataTest.java` | Add `testIntegerRowKeySerialization`, `testBigintRowKeySerialization`, `testSmallintRowKeySerialization`, `testTinyintRowKeySerialization`, `testExtractRowKeyAsString`, `testExtractRowKeyAsStringUnsupportedType` |

### Docs

| File | Change |
|---|---|
| `README.md` | Update Row Key section to list all supported types and add numeric PK example |
| `RowDataToRowMutationSerializer.java` | Update `@param rowKeyField` Javadoc |

## Backward compatibility

- Existing pipelines with `VARCHAR`/`CHAR` primary keys are completely unchanged — same code path, same validation
- No new configuration options — the feature is schema-driven
- No new dependencies

## Tophatting

- `./mvnw clean package -f connectors/bigtable/flink-connector-gcp-bigtable/pom.xml`
image

### CODE REVIEW

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java [hunk 1]
```diff
@@ -32,7 +32,7 @@ public class ErrorMessages {
     public static final String NESTED_TYPE_ERROR =
             "Nested rows are only supported with withNestedRowsMode and not double nested. Use Bytes for more complex types";
     public static final String ROW_KEY_STRING_TYPE_TEMPLATE =
-            "Row Key has to be of type String, got %s.";
+            "Row Key has to be of type STRING, BIGINT, INTEGER, SMALLINT, or TINYINT, got %s.";
```
Comment (brandtnewton) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java]: Should these types be the same as LogicalTypeRoot? i.e. "VARCHAR" instead of "STRING" and adding CHAR?

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java [hunk 2]
```diff
@@ -44,6 +45,16 @@
  * Sink for Table API.
  */
 public class BigtableDynamicTableSink implements DynamicTableSink {
+    /** Row key types that can be automatically converted to a String row key. */
+    static final EnumSet<LogicalTypeRoot> SUPPORTED_ROW_KEY_TYPES =
```
Comment (brandtnewton) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java]: nit: make the SUPPORTED_ROW_KEY_TYPES reusable

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataToRowMutationSerializer.java [hunk 3]
```diff
@@ -137,6 +151,37 @@ public RowMutationEntry serialize(RowData record, SinkWriter.Context context) {
...
+     * <p>Supports VARCHAR, CHAR, BIGINT, INTEGER, SMALLINT, and TINYINT types. Numeric types are
+     * converted using their standard {@code String.valueOf} representation.
+     *
+     * @param record The row data containing the key field.
+     * @param index The field index of the row key.
+     * @param typeRoot The logical type root of the row key field.
+     * @return The row key as a String.
+     */
+    @VisibleForTesting
+    static String extractRowKeyAsString(RowData record, int index, LogicalTypeRoot typeRoot) {
```
Comment (brandtnewton) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataToRowMutationSerializer.java:203]: do we really want to serialize numeric row key types as strings? This will break the lexicographic sorting of row keys (example: "10" sort before "2").

**Review by brandtnewton — COMMENTED** [review]
Thanks for the PR! Just had a couple of questions

**Comment by ma-g-22** [comment]
Thanks for the PR!

Could you please update your commit message/PR title to use the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) structure (we use this format to generate changelogs, etc)

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java (hunk 1 above)
Comment (wiliam-oliveira) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java]: Thanks! Fixed. Updated the error template to use VARCHAR and CHAR to match the LogicalTypeRoot enum names.

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java (hunk 2 above)
Comment (wiliam-oliveira) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java]: Fixed. Removed the duplicate definition from `BigtableDynamicTableSink` and made `SUPPORTED_ROW_KEY_TYPES` public in `RowDataToRowMutationSerializer`, which `BigtableDynamicTableSink` now references via static import.

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataToRowMutationSerializer.java (hunk 3 above)
Comment (wiliam-oliveira) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataToRowMutationSerializer.java:203]: @brandtnewton Good catch. Thank you!

The Bigtable schema design guide actually recommends this approach: https://docs.cloud.google.com/bigtable/docs/schema-design#general-concepts:~:text=If%20your%20data,queries%20are%20used.

Updated to zero-pad to 19 digits which is the maximum digit count for a BIGINT/Long. This preserves lexicographic sort order for non-negative values. Negative values are documented as not recommended for row keys since their sort order would be inverted.

We are also working on a separate PR to make row key formatting more extensible, for example, supporting composite row keys such as field_1#field_2 where each component can also be zero-padded. The zero-padding approach chosen here is intentionally consistent with that direction.

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java (hunk 1 above)
Comment (ma-g-22) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java]: Thanks! 

Should the row key types listed in `connectors/bigtable/README.md` be updated too?

Comment (wiliam-oliveira) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java]: Done!

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java [hunk 4]
```diff
@@ -32,7 +32,7 @@ public class ErrorMessages {
     public static final String NESTED_TYPE_ERROR =
             "Nested rows are only supported with withNestedRowsMode and not double nested. Use Bytes for more complex types";
     public static final String ROW_KEY_STRING_TYPE_TEMPLATE =
```
Comment (ma-g-22) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java]: Consider renaming this variable to `ROW_KEY_UNSUPPORTED_TYPE_TEMPLATE` (or similar), since it's not just String types now

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataToRowMutationSerializer.java [hunk 5]
```diff
@@ -169,6 +184,39 @@ public RowMutationEntry serialize(RowData record, SinkWriter.Context context) {
...
+            case INTEGER:
+                return String.format("%019d", record.getInt(index));
+            case SMALLINT:
+                return String.format("%019d", record.getShort(index));
+            case TINYINT:
+                return String.format("%019d", record.getByte(index));
+            default:
+                throw new IllegalArgumentException(
+                        String.format(ErrorMessages.ROW_KEY_STRING_TYPE_TEMPLATE, typeRoot));
+        }
```
[the remark this answers is not in the document]
Comment (wiliam-oliveira) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataToRowMutationSerializer.java:219]: Fixed, extracted into a protected static final String ROW_KEY_NUMERIC_FORMAT = "%019d" constant, with a comment noting that 19 is the number of digits in Long.MAX_VALUE

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java [hunk 6]
```diff
@@ -58,12 +59,15 @@ public BigtableDynamicTableSink(
...
-        checkArgument(
+        LogicalTypeRoot rowKeyType =
                 resolvedSchema
                         .getColumn(rowKeyIndex)
                         .get()
                         .getDataType()
-                        .equals(DataTypes.STRING().notNull()),
+                        .getLogicalType()
+                        .getTypeRoot();
+        checkArgument(
```
Comment (ma-g-22) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:66]: The original code performed a `.notNull()` check. Consider also checking that here?

File: connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataTest.java [hunk 7]
```diff
@@ -565,6 +567,137 @@ private static GenericRowData createNestedRow() {
...
+        RowMutationEntry entry = serializer.serialize(row, null);
+        assertEquals("0000000000000000007", entry.toProto().getRowKey().toStringUtf8());
+    }
+
+    @Test
+    public void testExtractRowKeyAsString() {
+        GenericRowData row = new GenericRowData(5);
+        row.setField(0, StringData.fromString("key1"));
+        row.setField(1, 42);
+        row.setField(2, 9876543210L);
```
Comment (ma-g-22) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataTest.java:664]: Consider adding a couple more test cases for edge case values such as `Long.MAX_VALUE`

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java (hunk 4 above)
Comment (ma-g-22) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java]: ( @wiliam-oliveira )

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java (hunk 6 above)
Comment (wiliam-oliveira) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:66]: Done!

File: connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataTest.java (hunk 7 above)
Comment (wiliam-oliveira) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataTest.java:664]: Added `testExtractRowKeyAsStringEdgeCases`

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java (hunk 4 above)
Comment (wiliam-oliveira) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/utils/ErrorMessages.java]: Done!

**Review by ma-g-22 — APPROVED** [review]
